### PR TITLE
kubeadm: improve resiliency when conflicts arise when updating the kubeadm-config configmap

### LINKED ISSUE #1097: Use configmap locks on kubeadm-config before the update.
https://github.com/kubernetes/kubernetes/pull/67944 changed how kubeadm uses the `kubeadm-config`ConfigMap, that now gets updated at every `kubeadm join --control-plane`.

Therefore we should now implement configmap locks on kubeadm-config before the update.

rif [code](
https://github.com/kubernetes/kubernetes/blob/e5f55dd9d0969af3a37e7880d9b88c9aef8c654d/cmd/kubeadm/app/phases/uploadconfig/uploadconfig.go#L61)

### PR DESCRIPTION
**Pull request by ereslibre**
**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:
Add the functionality to support `CreateOrMutateConfigMap` and `MutateConfigMap`.

* `CreateOrMutateConfigMap` will try to create a given config map object; if this configmap
  already exists, a new version of the resource will be retrieved from the server and a
  mutator callback will be called on it. Then, an `Update` of the mutated object will be
  performed. If there's a conflict during this `Update` operation, retry until no conflict
  happens. On every retry the object is refreshed from the server to the latest version.

* `MutateConfigMap` will try to get the latest version of the configmap from the server,
  call the mutator callback and then try to `Update` the mutated object. If there's a
  conflict during this `Update` operation, retry until no conlfict happens. On every retry
  the object is refreshed from the server to the latest version.

**Which issue(s) this PR fixes**:
Fixes https://github.com/kubernetes/kubeadm/issues/1097

**Special notes for your reviewer**:
Alternative implementation to https://github.com/kubernetes/kubernetes/pull/76241, without the need to perform leader election.

**Does this PR introduce a user-facing change?**:
```release-note
kubeadm: Improve resiliency when it comes to updating the `kubeadm-config` config map upon new control plane joins or resets. This allows for safe multiple control plane joins and/or resets.
```

### CODE REVIEW

File: cmd/kubeadm/app/util/apiclient/idempotency.go
Comment (neolit123) [inline cmd/kubeadm/app/util/apiclient/idempotency.go]: would it work if  we move the `cm` variable outside of the loop?

File: cmd/kubeadm/app/util/apiclient/idempotency.go
Comment (neolit123) [inline cmd/kubeadm/app/util/apiclient/idempotency.go]: 1) what would be the retry rate i.e. how many calls per second?
does it make sense to add a small "wait" in the for loop?

2) do we want to cap the number of retries to a relatively high number?

**Review by neolit123 — COMMENTED** [review]
thanks a lot for your research on this topic @ereslibre 
it seems so much better to not use LE and locks for this.

my guess would be that we should add a unit test for this too?

File: cmd/kubeadm/app/phases/uploadconfig/uploadconfig.go
Comment (neolit123) [inline cmd/kubeadm/app/phases/uploadconfig/uploadconfig.go]: `...APIEndpoint for Node...`

File: cmd/kubeadm/app/util/apiclient/idempotency.go
Comment (neolit123) [inline cmd/kubeadm/app/util/apiclient/idempotency.go]: `...create ConfigMap...`

File: cmd/kubeadm/app/util/apiclient/idempotency.go
Comment (neolit123) [inline cmd/kubeadm/app/util/apiclient/idempotency.go]: `...mutate ConfigMap`

Comment (neolit123) [inline cmd/kubeadm/app/util/apiclient/idempotency.go]: `...mutate ConfigMap`

File: cmd/kubeadm/app/util/apiclient/idempotency.go
Comment (neolit123) [reply cmd/kubeadm/app/util/apiclient/idempotency.go]: `configmap -> ConfigMap`

File: cmd/kubeadm/app/util/apiclient/idempotency.go
Comment (neolit123) [inline cmd/kubeadm/app/util/apiclient/idempotency.go]: `configmap -> ConfigMap`

**Comment by neolit123** [comment]
/priority important-soon
@kubernetes/sig-cluster-lifecycle-pr-reviews 

@ereslibre 
we should add a release note here too.

File: cmd/kubeadm/app/util/apiclient/idempotency.go
Comment (rosti) [inline cmd/kubeadm/app/util/apiclient/idempotency.go]: `cmmeta` and `cmmutator` are not very readable to me. `meta` and `mutator` seem good enough. Since the func is called `MutateConfigMap` the context is clear to that these are the config map meta and config map mutator.

File: cmd/kubeadm/app/util/apiclient/idempotency.go
Comment (rosti) [reply cmd/kubeadm/app/util/apiclient/idempotency.go]: Let's keep the var here, but change it to `configMap`.

File: cmd/kubeadm/app/util/apiclient/idempotency.go
Comment (detiber) [reply cmd/kubeadm/app/util/apiclient/idempotency.go]: Alternatively RetryOnConflict from client-go/util/retry could be used here: https://github.com/kubernetes/client-go/blob/master/util/retry/util.go#L61

Comment (rosti) [reply cmd/kubeadm/app/util/apiclient/idempotency.go]: Ideally we want to use `PollImmediateInfinite` here. Due to the nature of how the API server works and how we use the config map, this will end at some point (when all other clients have the CM updated). However we want an interval value in order to avoid bringing down an API server upon large number of simultaneous updates.

**Review by rosti — COMMENTED** [review]
Thanks @ereslibre !
I rather like this way better than the leader election method.

File: cmd/kubeadm/app/phases/uploadconfig/uploadconfig.go
Comment (timothysc) [inline cmd/kubeadm/app/phases/uploadconfig/uploadconfig.go]: This seems overly complicated form of what is typically referred to as CreateOrUpdate, I don't think it needs to be a closure at this level

File: cmd/kubeadm/app/phases/uploadconfig/uploadconfig.go
Comment (timothysc) [inline cmd/kubeadm/app/phases/uploadconfig/uploadconfig.go]: CreateOrUpdate

File: cmd/kubeadm/app/util/apiclient/idempotency.go
Comment (timothysc) [inline cmd/kubeadm/app/util/apiclient/idempotency.go]: CreateOrUpdate

File: cmd/kubeadm/app/util/apiclient/idempotency.go
Comment (timothysc) [inline cmd/kubeadm/app/util/apiclient/idempotency.go]: UpdateConfigMap

File: cmd/kubeadm/app/util/apiclient/idempotency.go
Comment (timothysc) [inline cmd/kubeadm/app/util/apiclient/idempotency.go]: What are the max retries and timeout on this?

**Review by timothysc — CHANGES_REQUESTED** [review]

File: cmd/kubeadm/app/phases/uploadconfig/uploadconfig.go
Comment (ereslibre) [reply cmd/kubeadm/app/phases/uploadconfig/uploadconfig.go]: The problem with `CreateOrUpdate` is that it involves some boilerplate that needs repeating both in reset and when joining new control plane nodes (unmarshal cluster status, mutate, marshal cluster status back). Also, the whole repeat if conflict would need to be repeated in both places.

`CreateOrMutateConfigMap` and `MutateConfigMap` approach this in a different way, you just implement the mutator and `MutateConfigMap` will take care of:

1. Fetch the resource
2. Callback to the mutator
3. Update
  3.1. If conflict, go to 1

Comment (ereslibre) [reply cmd/kubeadm/app/phases/uploadconfig/uploadconfig.go]: However, I will look into `RetryOnConflict` as @detiber suggested, it's exactly the same functionality that I wrote here.

File: cmd/kubeadm/app/util/apiclient/idempotency.go
Comment (ereslibre) [reply cmd/kubeadm/app/util/apiclient/idempotency.go]: @neolit123 we are not consistent on `configmap` vs `ConfigMap` on `idempotency.go`. Will update to `ConfigMap` what this patch is changing, but we probably want to follow up to make everything consistent.

File: cmd/kubeadm/app/phases/uploadconfig/uploadconfig.go
Comment (ereslibre) [reply cmd/kubeadm/app/phases/uploadconfig/uploadconfig.go]: Same as [1]

File: cmd/kubeadm/app/util/apiclient/idempotency.go
Comment (ereslibre) [reply cmd/kubeadm/app/util/apiclient/idempotency.go]: Same as [1]

File: cmd/kubeadm/app/util/apiclient/idempotency.go
Comment (ereslibre) [reply cmd/kubeadm/app/util/apiclient/idempotency.go]: Same as [1]

File: cmd/kubeadm/app/util/apiclient/idempotency.go
Comment (ereslibre) [reply cmd/kubeadm/app/util/apiclient/idempotency.go]: Updated to use `k8s.io/client-go/util/retry`

Comment (neolit123) [reply cmd/kubeadm/app/util/apiclient/idempotency.go]: we might want to use `Steps: 0` but with a sensibly large cap e.g. `Cap: 1 * time.Minute`:
https://godoc.org/k8s.io/apimachinery/pkg/util/wait#Backoff

Comment (detiber) [reply cmd/kubeadm/app/util/apiclient/idempotency.go]: I'd argue that someone adding 100 control plane nodes is doing something massively wrong.

File: cmd/kubeadm/app/util/apiclient/idempotency.go
Comment (timothysc) [inline cmd/kubeadm/app/util/apiclient/idempotency.go]: Don't you want to still pass if it's empty?

File: cmd/kubeadm/app/phases/uploadconfig/uploadconfig.go
Comment (timothysc) [reply cmd/kubeadm/app/phases/uploadconfig/uploadconfig.go]: k, I see why you're doing it this way now.  Thanks for the explanation.

File: cmd/kubeadm/app/util/apiclient/idempotency.go
Comment (timothysc) [inline cmd/kubeadm/app/util/apiclient/idempotency.go]: How are you ensuring multiple concurrent writers aren't overwriting each other?

**Review by timothysc — CHANGES_REQUESTED** [review]
Definitely needs a multiple concurrent writer test.

File: cmd/kubeadm/app/util/apiclient/idempotency.go
Comment (timothysc) [reply cmd/kubeadm/app/util/apiclient/idempotency.go]: Only a Patch would be safe in this scenario.

Comment (ereslibre) [reply cmd/kubeadm/app/util/apiclient/idempotency.go]: * [client1] Fetch `kubeadm-config` (v1)
* [client2] Fetch `kubeadm-config` (v1)
* [client1] Mutate `kubeadm-config` locally (v1)
* [client2] Mutate `kubeadm-config` locally (v1)
* [client1] Update `kubeadm-config` locally (v2)
* [client2] Update `kubeadm-config` locally (v1) - CONFLICT
* [client2] Fetch `kubeadm-config`  (v2)
* [client2] Mutate `kubeadm-config` locally (v2)
* [client2] Update `kubeadm-config` locally (v3)

`Patch` vs `Update` does not give more or less safety in this case, the conflict will happen in both cases, `Patch` is just handier to only provide a subset of properties to be altered. At least this is my understanding, please tell me if I'm understanding something in a wrong way.

File: cmd/kubeadm/app/util/apiclient/idempotency.go
Comment (ereslibre) [reply cmd/kubeadm/app/util/apiclient/idempotency.go]: `MutateConfigMap` has a slightly different semantics in my opinion than `CreateOrMutateConfigMap`; this is used when `reset` for example. I think that in this case it makes sense that we assume the configmap should exist (despite `reset` will ignore that error in any case because it wants to reset as much as possible, and not quit if the configmap is not present)

File: cmd/kubeadm/app/util/apiclient/idempotency.go
Comment (timothysc) [reply cmd/kubeadm/app/util/apiclient/idempotency.go]: Update doesn't care though it's blind and dump, there is no conflict, it's just a raw put. 
```
result = &v1.ConfigMap{}
        err = c.client.Put().
                Namespace(c.ns).
                Resource("configmaps").
                Name(configMap.Name).
                Body(configMap).
                Do().
                Into(result)
```
There is no conflict it's just a straight overwrite.

Comment (timothysc) [reply cmd/kubeadm/app/util/apiclient/idempotency.go]: Patch would indeed give you more safety b/c order would not matter.

Comment (ereslibre) [reply cmd/kubeadm/app/util/apiclient/idempotency.go]: I can ensure `Update` does indeed return a conflict if the version of the resource that you are updating is not the latest stored in `etcd`.

Comment (timothysc) [reply cmd/kubeadm/app/util/apiclient/idempotency.go]: So long as resource version is decoded from the `get` and passed again on the update you're right.  Sorry, it's been too long since I thought about this.

**Review by timothysc — APPROVED** [review]
/approve 

lgtm on tests.

**Comment by ereslibre** [comment]
/hold

I want to evaluate adding some tests and potentially add a release note as well.

File: cmd/kubeadm/app/util/apiclient/idempotency.go
Comment (ereslibre) [reply cmd/kubeadm/app/util/apiclient/idempotency.go]: @neolit123 `Steps: 0` won't work because of https://github.com/kubernetes/apimachinery/blob/b874eabb9a4eb99cef27db5c8d06f16542580cec/pkg/util/wait/wait.go#L283, the condition will never be executed.

**Comment by ereslibre** [comment]
Tests added: one just checks that `MutateConfigMap` acts as desired in a regular environment, whereas the other mimics 5 consecutive update conflicts, making the sixth update work.

PTAL and tell me, I've considered using a table driven test but thought this way was more clear, please provide feedback, if required I can update them.

I have also adapted the backoff a bit, it's mostly guessing because we'll probably need experience from the field to improve these backoff timings.

File: cmd/kubeadm/app/util/apiclient/idempotency.go
Comment (neolit123) [reply cmd/kubeadm/app/util/apiclient/idempotency.go]: i guess that's true. in that case we need to see what number of steps and cap we need to settle on. what is the retry rate you are seeing, btw?

Comment (ereslibre) [reply cmd/kubeadm/app/util/apiclient/idempotency.go]: @neolit123 all my tests with this are very artificial. I can try to do some real testing on this with e.g. 1 master + 4 concurrent joins, however it's also tricky to force the configmap update race on a complete join process, and running `kubeadm config upload from-file` again seems artificial to me.

I think it makes sense to start with some defaults that we consider sane and learn with the experience, sorry but I don't have a better answer since I believe giving a good enough answer requires some explicit efforts here.

Comment (neolit123) [reply cmd/kubeadm/app/util/apiclient/idempotency.go]: i think the artificial tests are OK as a base for the estimate here.
after all, we are only interested in the config map update.

so if a unit test passes for a excess number of concurrent clients, the real world case should be good too. at least that;'s how i see it.

we might be able to test this in e2e eventually too.

Comment (ereslibre) [reply cmd/kubeadm/app/util/apiclient/idempotency.go]: > so if a unit test passes for a excess number of concurrent clients, the real world case should be good too. at least that;'s how i see it.

Oh, you propose to start an apiserver for the unit tests? I have been using the fake client (that cannot mimic real server-side conflicts, it's just reactors that can mimic a conflict returned by the server, but it's not testing any real concurrency). I can look into that, but as far as I remember starting an apiserver was flaky and in general not desired, I might be wrong though.

Comment (neolit123) [reply cmd/kubeadm/app/util/apiclient/idempotency.go]: yeah, a real server is bad idea. it flakes in parallel unit tests; there are also endpoint and port conflicts etc. i was hopping that there is a fake api server that can be used here..

alternatively, we can have this as a test in test/e2e-kubeadm where a set of concurrent clients are trying to update the real ClusterStatus ConfigMap of the real server that the suite is being run on.

Comment (ereslibre) [reply cmd/kubeadm/app/util/apiclient/idempotency.go]: I think e2e tests are better for this purpose, because with a test server we would have to mimic ourselves a basic conflict return (version handling and so on); all in memory, but still, I think it will include some boilerplate code that we'll have to maintain in the end.

I will look into the e2e tests and see if they can be improved to include a multi-writer test. Do you want me to tackle that as part of this PR as well?

Comment (neolit123) [reply cmd/kubeadm/app/util/apiclient/idempotency.go]: thanks. i think a separate PR is better.

@fabriziopandini merged a change recently that allows you to run this kubeadm suite on an existing cluster like so: `make test-e2e-kubeadm`
https://github.com/kubernetes/kubernetes/pull/75546

Comment (ereslibre) [reply cmd/kubeadm/app/util/apiclient/idempotency.go]: I will work on that as a separate PR as soon as this one merges. Thank you!

**Review by rosti — COMMENTED** [review]
Thanks @ereslibre ! Looks great!
I have a couple of suggestions for additional test cases, but I think, that you have covered the most important ones.
What table testing gives us is the easier ability to add new test cases for the func at hand (provided, that the "table" was abstracted properly). I am fine with the way you chose.
Once you are happy with the tests, please remove the hold and ping us for final reviews.

**Comment by ereslibre** [comment]
> Once you are happy with the tests, please remove the hold and ping us for final reviews.

Thank you @rosti, I will add the tests that you propose and also start to work in parallel in the e2e tests, so maybe I can adapt this PR with @neolit123 concern about the backoff defaults when I gather some info from the e2e tests themselves. I have no strong opinions on whether this should merge as is (when the tests your mention are added) and then the backoff adapted after checking the e2e behavior, or if work on both in parallel so the backoff gets fixed on this very same PR with our findings.

**Comment by ereslibre** [comment]
@kubernetes/sig-cluster-lifecycle-pr-reviews this is ready for a final pass.

I think the default backoff in this PR is reasonable for this to merge directly, but while I'm working on the e2e tests I will update this PR with my latest findings if it's still open.

**Comment by ereslibre** [comment]
Also, I believe this patch is safe enough to be backported to 1.14, what do you think? We have some use cases for 1.14 that are interested in multi master concurrent joins.

cc/ @stealthybox

**Review by neolit123 — COMMENTED** [review]
i have minimal nitpicks about copy editing and code formatting, but nothing outstanding.
wait.Backoff LGTM
thanks for the update @ereslibre

File: cmd/kubeadm/app/phases/uploadconfig/uploadconfig_test.go
Comment (rosti) [inline cmd/kubeadm/app/phases/uploadconfig/uploadconfig_test.go]: Dump err here?

File: cmd/kubeadm/app/phases/uploadconfig/uploadconfig_test.go
Comment (rosti) [inline cmd/kubeadm/app/phases/uploadconfig/uploadconfig_test.go]: Dump err here?

File: cmd/kubeadm/app/phases/uploadconfig/uploadconfig_test.go
Comment (rosti) [inline cmd/kubeadm/app/phases/uploadconfig/uploadconfig_test.go]: s/Fatalf/Fatal, or even better, you can dump "expected" and "got" with `%v`.

File: cmd/kubeadm/app/util/apiclient/idempotency_test.go
Comment (neolit123) [inline cmd/kubeadm/app/util/apiclient/idempotency_test.go]: should be on a single line `const configMapName = ...`

File: cmd/kubeadm/app/util/apiclient/idempotency_test.go
Comment (rosti) [inline cmd/kubeadm/app/util/apiclient/idempotency_test.go]: s/Fatalf/Fatal

Comment (neolit123) [inline cmd/kubeadm/app/util/apiclient/idempotency_test.go]: i would use "ConfigMap" in all instances instead of "configmap" or "config map" even if the existing code already uses the later syntax.
[1]

File: cmd/kubeadm/app/util/apiclient/idempotency_test.go
Comment (neolit123) [inline cmd/kubeadm/app/util/apiclient/idempotency_test.go]: [1]

File: cmd/kubeadm/app/util/apiclient/idempotency_test.go
Comment (neolit123) [inline cmd/kubeadm/app/util/apiclient/idempotency_test.go]: [1]

File: cmd/kubeadm/app/util/apiclient/idempotency_test.go
Comment (neolit123) [inline cmd/kubeadm/app/util/apiclient/idempotency_test.go]: [1]

**Review by rosti — COMMENTED** [review]
Thanks @ereslibre !
Only a few tiny nits from me and it's otherwise LGTM.

File: cmd/kubeadm/app/util/apiclient/idempotency_test.go
Comment (neolit123) [inline cmd/kubeadm/app/util/apiclient/idempotency_test.go]: [1]

File: cmd/kubeadm/app/util/apiclient/idempotency_test.go
Comment (neolit123) [inline cmd/kubeadm/app/util/apiclient/idempotency_test.go]: [1]

File: cmd/kubeadm/app/util/apiclient/idempotency_test.go
Comment (neolit123) [inline cmd/kubeadm/app/util/apiclient/idempotency_test.go]: [1]

File: cmd/kubeadm/app/util/apiclient/idempotency_test.go
Comment (neolit123) [inline cmd/kubeadm/app/util/apiclient/idempotency_test.go]: [1]

**Comment by rosti** [comment]
> Also, I believe this patch is safe enough to be backported to 1.14, what do you think? We have some use cases for 1.14 that are interested in multi master concurrent joins.

I am 90% sure too, but the remaining 10% are making me worried.
In general, my rule is that only things, that are marked as bug+critical-urgent should get backported. I think, that it's safer for this to be tested a bit with the 1.15 master and eventually be released then.
Also, people have lived with this problem for now and there are available workarounds.

Therefore, I am a bit hesitant to backporting.

**Comment by neolit123** [comment]
the criteria for a "critical fix" is hard to define.

the workaround in previous versions is to manually sync the join of members.
this makes it difficult for users, but they are certainly not blocked.

**Comment by ereslibre** [comment]
I think I addressed all your feedback, thanks!

**Comment by rosti** [comment]
Thanks again @ereslibre !
I think, that this is now ok to be merged.

/lgtm
/hold cancel

**Pull request merged**
